mm/kfence/kfence-test.c

```diff
@@ -0,0 +1,485 @@
+
+static void setup_test_cache(struct kunit *test, size_t size, void (*ctor)(void *))
+{
+	if (test->priv != TEST_PRIV_WANT_MEMCACHE) {
+		KUNIT_ASSERT_FALSE_MSG(test, ctor, "unexpected ctor?");
+		return;
+	}
+
+	/*
+	 * Use SLAB_NOLEAKTRACE to prevent merging with existing caches. Any
+	 * other flag in SLAB_NEVER_MERGE (except SLAB_TYPESAFE_BY_RCU, which
+	 * disables KFENCE) also works. Use SLAB_ACCOUNT to allocate via memcg,
+	 * if enabled.
+	 */
+	test_cache = kmem_cache_create("test", size, 1, SLAB_NOLEAKTRACE | SLAB_ACCOUNT, ctor);
+	KUNIT_ASSERT_TRUE_MSG(test, test_cache, "could not create cache");
+}
+
+/* Must always inline to match stack trace against caller. */
+static __always_inline void test_free(void *ptr)
+{
+	if (test_cache)
+		kmem_cache_free(test_cache, ptr);
+	else
+		kfree(ptr);
+}
+
+/* On which side the allocation and the closest guard page should be. */
+enum allocation_side {
+	ALLOCATE_ANY,
+	ALLOCATE_LEFT,
+	ALLOCATE_RIGHT,
+};
+
+/*
+ * Try to get a guarded allocation from KFENCE. Uses either kmalloc() or the
+ * current test_cache if set up.
+ */
+static void *test_alloc(struct kunit *test, size_t size, gfp_t gfp, enum allocation_side side)
+{
+	void *alloc;
+	unsigned long timeout;
+
+	kunit_info(test, "%s: size=%zu, gfp=%x, side=%s, cache=%i", __func__, size, gfp,
+		   side == ALLOCATE_ANY ? "any" : (side == ALLOCATE_LEFT ? "left" : "right"),
+		   !!test_cache);
+
+	/*
+	 * 10x the sample rate should be more than enough to ensure we get a
+	 * KFENCE allocation.
+	 */
+	timeout = jiffies + msecs_to_jiffies(10 * CONFIG_KFENCE_SAMPLE_RATE);
+	do {
+		if (test_cache)
+			alloc = kmem_cache_alloc(test_cache, gfp);
+		else
+			alloc = kmalloc(size, gfp);
+
+		if (is_kfence_addr(alloc)) {
+			if (side == ALLOCATE_ANY)
+				return alloc;
+			if (side == ALLOCATE_LEFT && IS_ALIGNED((unsigned long)alloc, PAGE_SIZE))
+				return alloc;
+			if (side == ALLOCATE_RIGHT && !IS_ALIGNED((unsigned long)alloc, PAGE_SIZE))
+				return alloc;
+		}
+
+		test_free(alloc);
+	} while (jiffies < timeout);
+
+	KUNIT_ASSERT_TRUE_MSG(test, false, "failed to allocate from KFENCE");
+	return NULL; /* Unreachable. */
+}
+
+static void test_out_of_bounds_read(struct kunit *test)
+{
+	const size_t size = 32;
+	struct expect_report expect = {
+		.type = KFENCE_ERROR_OOB,
+		.fn = test_out_of_bounds_read,
+	};
+	char *buf;
+
+	setup_test_cache(test, size, NULL);
+
+	/* Test both sides. */
+
+	buf = test_alloc(test, size, GFP_KERNEL, ALLOCATE_LEFT);
+	expect.addr = buf - 1;
+	READ_ONCE(*expect.addr);
+	KUNIT_EXPECT_TRUE(test, report_matches(&expect));
+	test_free(buf);
+
+	buf = test_alloc(test, size, GFP_KERNEL, ALLOCATE_RIGHT);
+	expect.addr = buf + size;
+	READ_ONCE(*expect.addr);
+	KUNIT_EXPECT_TRUE(test, report_matches(&expect));
+	test_free(buf);
+}
+
+static void test_use_after_free_read(struct kunit *test)
+{
+	const size_t size = 32;
+	struct expect_report expect = {
+		.type = KFENCE_ERROR_UAF,
+		.fn = test_use_after_free_read,
```

Comment:
I couldn't come up with a macro that could be use instead of re-typing the function name for every test. Guess you also tried that?

Reply:
I thought about it, but it won't buy much -- I think it's cleaner this way. And if we mistype the function name, the test will fail anyway.